### PR TITLE
[new] promtail pipeline:  Promtail Rate Limit stage #5048

### PR DESCRIPTION
**Pull request by liguozhong**
Promtail Rate Limit stage #5048
image

### CODE REVIEW

**Comment by liguozhong** [comment]
@Whyeasy ,rate limit stage

File: clients/pkg/logentry/stages/limit.go [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	rateLimiter *rate.Limiter
+	dropCount   *prometheus.CounterVec
+}
+
+func (m *limitStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+		for e := range in {
+			if !m.shouldDrop(e) {
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/limit.go:90]: What do you think about using backpressure instead? Something like
```golang
for {
    if m.rateLimiter.Allow() {
        e := <- in
        out <- e
    } else {
        // sleep
    }
}
```

**Review by jeschkies — COMMENTED** [review]
Thanks for proposing this change :slightly_smiling_face:

File: clients/pkg/logentry/stages/limit.go [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+		return false
+	}
+	if m.cfg.Drop {
+		if m.rateLimiter.Allow() {
+			return false
+		}
+		m.dropCount.WithLabelValues(ratelimitDropReason).Inc()
+		return true
+	}
+	_ = m.rateLimiter.Wait(context.Background())
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/limit.go:112]: This is awesome!

File: clients/pkg/logentry/stages/limit.go [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+			if !m.shouldDrop(e) {
+				out <- e
+				continue
+			}
+		}
+	}()
+	return out
+}
+
+func (m *limitStage) shouldDrop(e Entry) bool {
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/limit.go:100]: How about
```suggestion
func (m *limitStage) shouldThrottle(e Entry) bool {
```

**Review by jeschkies — COMMENTED** [review]
Thanks for responding so quickly. I wonder how we could add a test :thinking:

**Comment by liguozhong** [comment]
> Thanks for responding so quickly. I wonder how we could add a test 🤔
done. add TestLimitWaitPipeline test drop:false case ,and TestLimitDropPipeline test drop:true case
image
image

File: clients/pkg/logentry/stages/limit.go [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+
+type LimitConfig struct {
+	Source     *string `mapstructure:"source"`
+	Rate       float64 `mapstructure:"rate"`
+	Burst      int     `mapstructure:"burst"`
+	Drop       bool    `mapstructure:"drop"`
+	Value      *string `mapstructure:"value"`
+	Expression *string `mapstructure:"expression"`
+	regex      *regexp.Regexp
+}
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/limit.go:34]: Sorry, I should have spotted this early. Would you mind adding documentation in (https://github.com/grafana/loki/blob/main/docs/sources/clients/promtail/pipelines.md#stages) as well? I'm not sure I understand `rate` and `burst`.

File: clients/pkg/logentry/stages/limit_test.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+- limit:
+    source: app
+    value: loki
+    rate: 1
+    burst: 1
+    drop: true
+`
+
+// TestLimitPipeline is used to verify we properly parse the yaml config and create a working pipeline
+func TestLimitWaitPipeline(t *testing.T) {
```
Comment (jeschkies) [inline clients/pkg/logentry/stages/limit_test.go:45]: :pray:

**Review by jeschkies — APPROVED** [review]
Nice. thanks for your patience!

**Comment by liguozhong** [comment]
continuous-integration/drone/pr — Build is failing
why,how to solve？
image

**Comment by jeschkies** [comment]
> why,how to solve？

Probably a flaky test.

**Comment by liguozhong** [comment]
Today s3 is unavailable again for 4 hours。
image

File: docs/sources/clients/promtail/stages/limit.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
+---
+title: limit
+---
+# `limit` stage
+
+The `limit` stage is a ratelimit stage that lets you throttle logs based on several options.
```
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/stages/limit.md:6]:
```suggestion
The `limit` stage is a rate limiting stage throttles logs based on several options. 
```

File: docs/sources/clients/promtail/stages/limit.md [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+
+  # allowed ingestion rate limit in lines per second
+  [rate: <int>]
+
+  # allowed ingestion burst lines
+  [burst: <int>]
+
+  # When drop is equal to true, logs that exceed the current ratelimit will be discarded.
+  # When drop is equal to false, the log that exceeds the current ratelimit will only wait to enter the backpressure mode 
+  [drop: <bool> | default = false]
```
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/stages/limit.md:32]: I'm correcting grammar and punctuation. I'm also reformatting such that line lengths will be smaller. This prevents the rendered page from needing a horizontal scroll bar.

```suggestion
  # Name from the extracted data to parse. If empty, it parses the log message.
  [source: <string>]
  
  # An RE2 regular expression. If the source is provided, the regex will attempt to match
  # the source. If no source is provided, then the regex attempts to match the log line.
  # If the provided regex matches the log line or a provided source, the line will
  # be throttled. 
  [expression: <string>]

  # A value can only be specified when `source` is specified. It is an error to specify
  # both `value` and `expression`. If the value provided is an exact match for
  # the `source`, the line will be throttled.
  [value: <string>]

  # The allowed ingestion rate limit in lines per second
  [rate: <int>]

  # The quantity of allowed ingestion burst lines
  [burst: <int>]

  # When drop is true, log lines that exceed the current rate limit will be discarded.
  # When drop is false, log lines that exceed the current rate limit will only wait
  # to enter the back pressure mode. 
  [drop: <bool> | default = false]
```

**Review by KMiller-Grafana — CHANGES_REQUESTED** [review]
Every other stage provides one or more examples.  Can you provide one for this limit stage?

**Comment by liguozhong** [comment]
> > why,how to solve？
> 
> Probably a flaky test.

done,thanks review
image

File: clients/pkg/logentry/stages/limit.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+var ratelimitDropReason = "ratelimit_drop_stage"
+
+type LimitConfig struct {
+	Source     *string `mapstructure:"source"`
+	Rate       float64 `mapstructure:"rate"`
+	Burst      int     `mapstructure:"burst"`
+	Drop       bool    `mapstructure:"drop"`
+	Value      *string `mapstructure:"value"`
+	Expression *string `mapstructure:"expression"`
```
Comment (cyriltovena) [inline clients/pkg/logentry/stages/limit.go:32]: Why do we need Value, Expression and Source here ? We could easily combine this stage with a `match` stage no ? I wonder if we need this logic in the limit stage ? see https://grafana.com/docs/loki/latest/clients/promtail/stages/match/

```yaml
match:
  selector: '{app="foo"} |~ "regex"'
  stages:
  - limit:
      rate: 10
      drop: true
```

@grafana/loki-team WDYT ? The only difference is that you can't use an extracted value, but regex goes already a long way.

Comment (sandeepsukhani) [reply clients/pkg/logentry/stages/limit.go:32]: Yeah, I think using the existing `match` stage for this makes sense to me.

File: clients/pkg/logentry/stages/limit.go [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+						level.Debug(m.logger).Log("msg", "line met drop criteria for finding source key in extracted map with value matching desired drop value")
+					}
+				} else {
+					// Value doesn't match, don't drop
+					if Debug {
+						level.Debug(m.logger).Log("msg", fmt.Sprintf("line will not be dropped, source key was found in extracted map but value '%v' did not match desired value '%v'", v, *m.cfg.Value))
+					}
+					return false
+				}
+			}
```
Comment (dannykopping) [inline clients/pkg/logentry/stages/limit.go:142]: This is pretty hard to read, with the multiple levels of indentation.

One thing that'd make it far simpler is having the ability to just run `level.Debug(m.logger).Log(...)` without checking `if Debug` first.

File: clients/pkg/logentry/stages/limit.go [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+
+func (m *limitStage) Run(in chan Entry) chan Entry {
+	out := make(chan Entry)
+	go func() {
+		defer close(out)
+		for e := range in {
+			if !m.shouldThrottle(e) {
+				out <- e
+				continue
+			}
```
Comment (chaudum) [inline clients/pkg/logentry/stages/limit.go:94]: I would not expect any side-effects from a function named `shouldThrottle` but only an evaluation whether throttling needs to be applied. What do you think about splitting the check and the actual throttling logic into separate functions?

File: clients/pkg/logentry/stages/limit.go (hunk 8 above)
Comment (cyriltovena) [reply clients/pkg/logentry/stages/limit.go:32]: Ok let's remove all this logic, if we have a request later we can think about how to do that. I want to keep each stage responsibility small and simple.

**Comment by liguozhong** [comment]
cyriltovena 1 hour ago
Ok let's remove all this logic, if we have a request later we can think about how to do that. I want to keep each stage responsibility small and simple.

done.

File: docs/sources/clients/promtail/stages/limit.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+
+```yaml
+- limit:
+    rate: 10
+    burst: 10
+```
+
+Would throttle any log line.
+
+#### Regex match a line and drop
```
Comment (cyriltovena) [inline docs/sources/clients/promtail/stages/limit.md:44]:
```suggestion
#### Match a line and drop
```

**Review by cyriltovena — APPROVED** [review]
LGTM

**Comment by liguozhong** [comment]
continuous-integration/drone/pr — Build is failing
why？

image

**Comment by cyriltovena** [comment]
Restarting. it's a flake.

**Comment by cyriltovena** [comment]
@KMiller-Grafana We need your approval to merge this one.

Thank you !

File: docs/sources/clients/promtail/stages/limit.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+## Limit stage schema
+
+```yaml
+limit:
+  # The allowed ingestion rate limit in lines per second
+  [rate: <int>]
+
+  # The quantity of allowed ingestion burst lines
+  [burst: <int>]
```
Comment (KMiller-Grafana) [inline docs/sources/clients/promtail/stages/limit.md:16]:
```suggestion

This pipeline stage places limits on the rate or burst quantity of log lines that Promtail pushes to Loki.
The rate and burst limits are equivalent to the limits specified by the distributor configuration parameters  `ingestion_rate_mb` and `ingestion_burst_size_mb`, as defined in [limits_config](../../../../configuration/#limits_config).

```yaml
limit:
  # The rate limit in lines per second that Promtail will push to Loki
  [rate: <int>]

  # The cap in the quantity of burst lines that Promtail will push to Loki
  [burst: <int>]
```

**Review by KMiller-Grafana — CHANGES_REQUESTED** [review]
I'm suggesting changes that give a bit of general understanding of what these rate limits are about.  Specifically, they don't implement an ingestion rate limit within Promtail; they instead limit what Promtail pushes to Loki.

File: docs/sources/clients/promtail/stages/limit.md (hunk 12 above)
Comment (09jvilla) [reply docs/sources/clients/promtail/stages/limit.md:16]: I made an edit to the comment above so it now reads
```
This pipeline stage places limits on the rate or burst quantity of log lines that Promtail pushes to Loki.
The concept of having distinct burst and rate limits mirrors the approach to limits that can be set for Loki's distributor component:  `ingestion_rate_mb` and `ingestion_burst_size_mb`, as defined in [limits_config](../../../../configuration/#limits_config).

```

**Comment by liguozhong** [comment]
> I'm suggesting changes that give a bit of general understanding of what these rate limits are about. Specifically, they don't implement an ingestion rate limit within Promtail; they instead limit what Promtail pushes to Loki.

done，thanks for your help.

**Comment by liguozhong** [comment]
> @liguozhong -- We made a few suggested improvements to the documentation. Can you take a look and merge them in if they look good to you?
> 
> Then we can get this merged to main :)

done，thanks for your help.

**Pull request merged**